vespalib/src/vespa/vespalib/test/time_tracer.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "time_tracer.h"
+
+namespace vespalib::test {
+
+//-----------------------------------------------------------------------------
+
+TimeTracer &
+TimeTracer::master()
+{
+    static TimeTracer instance;
+    return instance;
+}
+
+//-----------------------------------------------------------------------------
+
+thread_local TimeTracer::ThreadState *TimeTracer::_thread_state = nullptr;
+
+//-----------------------------------------------------------------------------
+
+TimeTracer::Tag::Tag(const vespalib::string &name)
+    : _id(master().get_tag_id(name))
+{
+}
+
+TimeTracer::Tag::~Tag() = default;
+
+//-----------------------------------------------------------------------------
+
+void
+TimeTracer::init_thread_state() noexcept
+{
+    _thread_state = master().create_thread_state();
+}
+
+//-----------------------------------------------------------------------------
+
+TimeTracer::TimeTracer()
+    : _lock(),
+      _state_list(),
+      _tags()
+{
+}
+
+uint32_t
+TimeTracer::get_tag_id(const vespalib::string &tag_name)
+{
+    auto guard = std::lock_guard(_lock);
```

Comment:
Purely out of interest, why this instead of `std::lock_guard guard(_lock);`?

Reply:
fixed (did not remember the best way to enable auto-deduction)